### PR TITLE
Teach ts sdk about product utility members

### PR DESCRIPTION
**Pull request by DanielCollins**

### CODE REVIEW

File: typescript/src/entities/domain.ts [hunk 1]
```diff
@@ -124,4 +124,28 @@ export class Domain extends Entity {
 
   @Domain.property({arrayType: "Theme"})
   public themes?: Array<Theme>;
+
+  public defaultCurrency = () => {
+    if (this.company === undefined) {
+      throw new Error("company is undefined, did you forget to embed it?");
+    }
+    if (this.company.defaultCurrency === undefined) {
+      const err = "company.defaultCurrency is undefined, did you forget to" +
```
Comment (s3341458) [inline typescript/src/entities/domain.ts]: default currency is a primary variable how can it not be embeded

File: typescript/src/entities/product.ts [hunk 2]
```diff
@@ -139,4 +139,138 @@ export class Product extends Entity {
...
+    if (this.independentVariationFields === undefined) {
+      const err = "independentVariationFields is undefined, did you forget to" +
+        " embed it?";
+      throw new Error(err);
+    }
+    return this.independentVariationFields.length > 0;
+  }
+
+  public taxType = () => {
+    if (this.domain === undefined) {
```
Comment (s3341458) [inline typescript/src/entities/product.ts]: can wrap in a function checkDomainEmbeded()

File: typescript/src/entities/product.ts [hunk 3]
```diff
@@ -139,4 +139,138 @@ export class Product extends Entity {
...
+
+  public currency = () => {
+    if (this.domain === undefined) {
+      throw new Error("domain is undefined, did you forget to embed it?");
+    }
+    return this.domain.defaultCurrency();
+  }
+
+  public hasGroupVariationFields = () => {
+    if (this.groupVariationFields === undefined) {
```
Comment (s3341458) [inline typescript/src/entities/product.ts]: ideally can  be warp in a function checkEmbeded(this.groupVariationFields) or do some magic on property annotations.

Not required for this pr, but just raise this issue #241

File: typescript/src/entities/product.ts [hunk 4]
```diff
@@ -139,4 +139,138 @@ export class Product extends Entity {
...
+    }
+    if (this.groupVariationFields === undefined) {
+      const err = "groupVariationFields is undefined, did you forget to embed" +
+        " it?";
+      throw new Error(err);
+    }
+    const variationFields = variationField.independent ?
+      this.independentVariationFields : this.groupVariationFields;
+    const index = variationFields.findIndex(v => {
+      if (v.id === undefined) {
```
Comment (s3341458) [inline typescript/src/entities/product.ts]: how could a miss embed cause variation.id to be null.

File: typescript/src/entities/variation_field.ts [hunk 5]
```diff
@@ -88,4 +90,58 @@ export class VariationField extends Entity {
...
+    const selectable = new Set([FieldType.SELECT,
+                                FieldType.CHECKBOX,
+                                FieldType.RADIO,
+                                FieldType.IMAGE_SELECT,
+                                FieldType.COLOUR_SELECT]);
+    return selectable.has(this.fieldType);
+  }
+
+  public buildEmptyVariation = () => {
+    if (this.defaultValue === undefined) {
```
Comment (s3341458) [inline typescript/src/entities/variation_field.ts]: how could a embed missing lead to undefined "defaultValue"

File: typescript/src/entities/variation_field.ts [hunk 6]
```diff
@@ -88,4 +90,58 @@ export class VariationField extends Entity {
...
+      throw new Error("defaultValue is undefined, did you forget to embed it?");
+    }
+    if (this.variationCost === undefined) {
```
Comment (s3341458) [inline typescript/src/entities/variation_field.ts]: how could a embed missing lead to undefined "variationCost"

File: typescript/src/entities/variation_field.ts [hunk 7]
```diff
@@ -88,4 +90,58 @@ export class VariationField extends Entity {
...
+      throw new Error("options is undefined, did you forget to embed it?");
+    }
+    const result = new this.merchi.Variation(this.merchi);
+    if (this.isSelectable()) {
+      let onceOffCost = 0;
+      const value = [];
+      for (const option of this.options) {
+        if (option.default) {
+          if (option.variationCost === undefined) {
+            throw new Error("option.variationCost is undefined, did you " +
```
Comment (s3341458) [inline typescript/src/entities/variation_field.ts]: same suspection why do embed check on primary value

Comment (DanielCollins) [reply typescript/src/entities/variation_field.ts]: OK, the error suggestion is bad, it's unlikely to be from a missed embed.

the situation will more likely arise due creating a new object that hasn't fetched anything at all from the server yet.

**Comment by s3341458** [comment]
as currently our sdk is only used internally, I think the error does not matter that much as long as we know what we are doing

File: typescript/src/entities/product.ts (hunk 2 above)
Comment (DanielCollins) [reply typescript/src/entities/product.ts]: not until typescript supports "assertion signatures" (i.e., 3.7)

**Pull request merged**